# 6.1 Sourcing open data

## This script contains the following:

### 1. Import libraries

### 2. Import data set

### 3. Data Quality

## 01. Import libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

## 02. Import data

In [2]:
# path folder
path = r'C:\Users\alexp\UFC Stats'
path

'C:\\Users\\alexp\\UFC Stats'

In [3]:
# import UFC data
df_ufc = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'data.csv'), index_col = False)

In [4]:
print(pd.options.display.max_columns)

20


In [5]:
# where 150 is the maximum number of columns that you want to show
pd.set_option('display.max_columns', 150)

In [6]:
# print first 5 rows of the df_ufc dataframe
df_ufc.head()

R_fighter          B_fighter        Referee        date  \
0   Adrian Yanez      Gustavo Lopez  Chris Tognoni  2021-03-20   
1   Trevin Giles      Roman Dolidze      Herb Dean  2021-03-20   
2    Tai Tuivasa    Harry Hunsucker      Herb Dean  2021-03-20   
3  Cheyanne Buys  Montserrat Conejo     Mark Smith  2021-03-20   
4  Marion Reneau      Macy Chiasson     Mark Smith  2021-03-20   

                 location Winner  title_bout       weight_class  B_avg_KD  \
0  Las Vegas, Nevada, USA    Red       False       Bantamweight     0.000   
1  Las Vegas, Nevada, USA    Red       False       Middleweight     0.500   
2  Las Vegas, Nevada, USA    Red       False        Heavyweight       NaN   
3  Las Vegas, Nevada, USA   Blue       False   WomenStrawweight       NaN   
4  Las Vegas, Nevada, USA   Blue       False  WomenBantamweight     0.125   

   B_avg_opp_KD  B_avg_SIG_STR_pct  B_avg_opp_SIG_STR_pct  B_avg_TD_pct  \
0           0.0           0.420000                0.49500         0.330   
1           0.0           0.660000                0.30500         0.300   
2           NaN                NaN                    NaN           NaN   
3           NaN                NaN                    NaN           NaN   
4           0.0           0.535625                0.57875         0.185   

   B_avg_opp_TD_pct  B_avg_SUB_ATT  B_avg_opp_SUB_ATT  B_avg_REV  \
0           0.36000          0.500             1.0000       0.00   
1           0.50000          1.500             0.0000       0.00   
2               NaN            NaN                NaN        NaN   
3               NaN            NaN                NaN        NaN   
4           0.16625          0.125             0.1875       0.25   

   B_avg_opp_REV  B_avg_SIG_STR_att  B_avg_SIG_STR_landed  \
0            0.0            50.0000               20.0000   
1            0.0            65.5000               35.0000   
2            NaN                NaN                   NaN   
3            NaN                NaN                   NaN   
4            0.0           109.1875               57.9375   

   B_avg_opp_SIG_STR_att  B_avg_opp_SIG_STR_landed  B_avg_TOTAL_STR_att  \
0                 84.000                   45.0000              76.5000   
1                 50.000                   16.5000             113.5000   
2                    NaN                       NaN                  NaN   
3                    NaN                       NaN                  NaN   
4                 50.625                   28.4375             170.4375   

   B_avg_TOTAL_STR_landed  B_avg_opp_TOTAL_STR_att  \
0                  41.000                  114.000   
1                  68.500                   68.500   
2                     NaN                      NaN   
3                     NaN                      NaN   
4                 105.625                   74.375   

   B_avg_opp_TOTAL_STR_landed  B_avg_TD_att  B_avg_TD_landed  \
0                       64.00         1.500              1.0   
1                       29.00         2.500              1.5   
2                         NaN           NaN              NaN   
3                         NaN           NaN              NaN   
4                       44.25         5.375              1.5   

   B_avg_opp_TD_att  B_avg_opp_TD_landed  B_avg_HEAD_att  B_avg_HEAD_landed  \
0               9.0                6.500          39.500             11.000   
1               0.5                0.500          46.000             20.000   
2               NaN                  NaN             NaN                NaN   
3               NaN                  NaN             NaN                NaN   
4               2.0                0.625          77.375             31.375   

   B_avg_opp_HEAD_att  B_avg_opp_HEAD_landed  B_avg_BODY_att  \
0             63.0000                 27.500          7.5000   
1             36.0000                  7.500         12.0000   
2                 NaN                    NaN             NaN   
3                 NaN       

In [7]:
# return tuple of number of rows and number of columns for df_ufc dataframe
df_ufc.shape

(6012, 144)

In [8]:
pd.set_option('display.max_rows', None)

In [9]:
# check datatypes of the df_ufc dataframe
df_ufc.dtypes

R_fighter                        object
B_fighter                        object
Referee                          object
date                             object
location                         object
Winner                           object
title_bout                         bool
weight_class                     object
B_avg_KD                        float64
B_avg_opp_KD                    float64
B_avg_SIG_STR_pct               float64
B_avg_opp_SIG_STR_pct           float64
B_avg_TD_pct                    float64
B_avg_opp_TD_pct                float64
B_avg_SUB_ATT                   float64
B_avg_opp_SUB_ATT               float64
B_avg_REV                       float64
B_avg_opp_REV                   float64
B_avg_SIG_STR_att               float64
B_avg_SIG_STR_landed            float64
B_avg_opp_SIG_STR_att           float64
B_avg_opp_SIG_STR_landed        float64
B_avg_TOTAL_STR_att             float64
B_avg_TOTAL_STR_landed          float64
B_avg_opp_TOTAL_STR_att         float64


## 03. Data Quality

### Data Wrangling

### 01. Dropping columns
### 02. Renaming columns
### 03. Changing data types
### 04. Transposing data

### Dropping columns

In [10]:
df_ufc.head()

R_fighter          B_fighter        Referee        date  \
0   Adrian Yanez      Gustavo Lopez  Chris Tognoni  2021-03-20   
1   Trevin Giles      Roman Dolidze      Herb Dean  2021-03-20   
2    Tai Tuivasa    Harry Hunsucker      Herb Dean  2021-03-20   
3  Cheyanne Buys  Montserrat Conejo     Mark Smith  2021-03-20   
4  Marion Reneau      Macy Chiasson     Mark Smith  2021-03-20   

                 location Winner  title_bout       weight_class  B_avg_KD  \
0  Las Vegas, Nevada, USA    Red       False       Bantamweight     0.000   
1  Las Vegas, Nevada, USA    Red       False       Middleweight     0.500   
2  Las Vegas, Nevada, USA    Red       False        Heavyweight       NaN   
3  Las Vegas, Nevada, USA   Blue       False   WomenStrawweight       NaN   
4  Las Vegas, Nevada, USA   Blue       False  WomenBantamweight     0.125   

   B_avg_opp_KD  B_avg_SIG_STR_pct  B_avg_opp_SIG_STR_pct  B_avg_TD_pct  \
0           0.0           0.420000                0.49500         0.330   
1           0.0           0.660000                0.30500         0.300   
2           NaN                NaN                    NaN           NaN   
3           NaN                NaN                    NaN           NaN   
4           0.0           0.535625                0.57875         0.185   

   B_avg_opp_TD_pct  B_avg_SUB_ATT  B_avg_opp_SUB_ATT  B_avg_REV  \
0           0.36000          0.500             1.0000       0.00   
1           0.50000          1.500             0.0000       0.00   
2               NaN            NaN                NaN        NaN   
3               NaN            NaN                NaN        NaN   
4           0.16625          0.125             0.1875       0.25   

   B_avg_opp_REV  B_avg_SIG_STR_att  B_avg_SIG_STR_landed  \
0            0.0            50.0000               20.0000   
1            0.0            65.5000               35.0000   
2            NaN                NaN                   NaN   
3            NaN                NaN                   NaN   
4            0.0           109.1875               57.9375   

   B_avg_opp_SIG_STR_att  B_avg_opp_SIG_STR_landed  B_avg_TOTAL_STR_att  \
0                 84.000                   45.0000              76.5000   
1                 50.000                   16.5000             113.5000   
2                    NaN                       NaN                  NaN   
3                    NaN                       NaN                  NaN   
4                 50.625                   28.4375             170.4375   

   B_avg_TOTAL_STR_landed  B_avg_opp_TOTAL_STR_att  \
0                  41.000                  114.000   
1                  68.500                   68.500   
2                     NaN                      NaN   
3                     NaN                      NaN   
4                 105.625                   74.375   

   B_avg_opp_TOTAL_STR_landed  B_avg_TD_att  B_avg_TD_landed  \
0                       64.00         1.500              1.0   
1                       29.00         2.500              1.5   
2                         NaN           NaN              NaN   
3                         NaN           NaN              NaN   
4                       44.25         5.375              1.5   

   B_avg_opp_TD_att  B_avg_opp_TD_landed  B_avg_HEAD_att  B_avg_HEAD_landed  \
0               9.0                6.500          39.500             11.000   
1               0.5                0.500          46.000             20.000   
2               NaN                  NaN             NaN                NaN   
3               NaN                  NaN             NaN                NaN   
4               2.0                0.625          77.375             31.375   

   B_avg_opp_HEAD_att  B_avg_opp_HEAD_landed  B_avg_BODY_att  \
0             63.0000                 27.500          7.5000   
1             36.0000                  7.500         12.0000   
2                 NaN                    NaN             NaN   
3                 NaN       

#### Referee column does not seem important for this project. Delete the column.

In [11]:
# Dropping referee column from df_ufc
df_ufc = df_ufc.drop(columns = ['Referee'])

In [12]:
# check updated df_ufc dataframe
df_ufc.head()

R_fighter          B_fighter        date                location  \
0   Adrian Yanez      Gustavo Lopez  2021-03-20  Las Vegas, Nevada, USA   
1   Trevin Giles      Roman Dolidze  2021-03-20  Las Vegas, Nevada, USA   
2    Tai Tuivasa    Harry Hunsucker  2021-03-20  Las Vegas, Nevada, USA   
3  Cheyanne Buys  Montserrat Conejo  2021-03-20  Las Vegas, Nevada, USA   
4  Marion Reneau      Macy Chiasson  2021-03-20  Las Vegas, Nevada, USA   

  Winner  title_bout       weight_class  B_avg_KD  B_avg_opp_KD  \
0    Red       False       Bantamweight     0.000           0.0   
1    Red       False       Middleweight     0.500           0.0   
2    Red       False        Heavyweight       NaN           NaN   
3   Blue       False   WomenStrawweight       NaN           NaN   
4   Blue       False  WomenBantamweight     0.125           0.0   

   B_avg_SIG_STR_pct  B_avg_opp_SIG_STR_pct  B_avg_TD_pct  B_avg_opp_TD_pct  \
0           0.420000                0.49500         0.330           0.36000   
1           0.660000                0.30500         0.300           0.50000   
2                NaN                    NaN           NaN               NaN   
3                NaN                    NaN           NaN               NaN   
4           0.535625                0.57875         0.185           0.16625   

   B_avg_SUB_ATT  B_avg_opp_SUB_ATT  B_avg_REV  B_avg_opp_REV  \
0          0.500             1.0000       0.00            0.0   
1          1.500             0.0000       0.00            0.0   
2            NaN                NaN        NaN            NaN   
3            NaN                NaN        NaN            NaN   
4          0.125             0.1875       0.25            0.0   

   B_avg_SIG_STR_att  B_avg_SIG_STR_landed  B_avg_opp_SIG_STR_att  \
0            50.0000               20.0000                 84.000   
1            65.5000               35.0000                 50.000   
2                NaN                   NaN                    NaN   
3                NaN                   NaN                    NaN   
4           109.1875               57.9375                 50.625   

   B_avg_opp_SIG_STR_landed  B_avg_TOTAL_STR_att  B_avg_TOTAL_STR_landed  \
0                   45.0000              76.5000                  41.000   
1                   16.5000             113.5000                  68.500   
2                       NaN                  NaN                     NaN   
3                       NaN                  NaN                     NaN   
4                   28.4375             170.4375                 105.625   

   B_avg_opp_TOTAL_STR_att  B_avg_opp_TOTAL_STR_landed  B_avg_TD_att  \
0                  114.000                       64.00         1.500   
1                   68.500                       29.00         2.500   
2                      NaN                         NaN           NaN   
3                      NaN                         NaN           NaN   
4                   74.375                       44.25         5.375   

   B_avg_TD_landed  B_avg_opp_TD_att  B_avg_opp_TD_landed  B_avg_HEAD_att  \
0              1.0               9.0                6.500          39.500   
1              1.5               0.5                0.500          46.000   
2              NaN               NaN                  NaN             NaN   
3              NaN               NaN                  NaN             NaN   
4              1.5               2.0                0.625          77.375   

   B_avg_HEAD_landed  B_avg_opp_HEAD_att  B_avg_opp_HEAD_landed  \
0             11.000             63.0000                 27.500   
1             20.000             36.0000                  7.500   
2                NaN                 NaN                    NaN   
3                NaN                 NaN                    NaN   
4             31.375             41.5625                 22.625   

   B_avg_BODY_att  B_avg_BODY_landed  B_avg_opp_BODY_att  \
0          7.5000             7.0000             12.0000

### Renaming columns

In [13]:
# renaming columns in the df_ufc dataframe
df_ufc.rename(columns = {'R_fighter' : 'red_fighter'}, inplace = True)
df_ufc.rename(columns = {'B_fighter' : 'blue_fighter'}, inplace = True)
df_ufc.rename(columns = {'Winner' : 'winner'}, inplace = True)

In [14]:
df_ufc.head()

red_fighter       blue_fighter        date                location  \
0   Adrian Yanez      Gustavo Lopez  2021-03-20  Las Vegas, Nevada, USA   
1   Trevin Giles      Roman Dolidze  2021-03-20  Las Vegas, Nevada, USA   
2    Tai Tuivasa    Harry Hunsucker  2021-03-20  Las Vegas, Nevada, USA   
3  Cheyanne Buys  Montserrat Conejo  2021-03-20  Las Vegas, Nevada, USA   
4  Marion Reneau      Macy Chiasson  2021-03-20  Las Vegas, Nevada, USA   

  winner  title_bout       weight_class  B_avg_KD  B_avg_opp_KD  \
0    Red       False       Bantamweight     0.000           0.0   
1    Red       False       Middleweight     0.500           0.0   
2    Red       False        Heavyweight       NaN           NaN   
3   Blue       False   WomenStrawweight       NaN           NaN   
4   Blue       False  WomenBantamweight     0.125           0.0   

   B_avg_SIG_STR_pct  B_avg_opp_SIG_STR_pct  B_avg_TD_pct  B_avg_opp_TD_pct  \
0           0.420000                0.49500         0.330           0.36000   
1           0.660000                0.30500         0.300           0.50000   
2                NaN                    NaN           NaN               NaN   
3                NaN                    NaN           NaN               NaN   
4           0.535625                0.57875         0.185           0.16625   

   B_avg_SUB_ATT  B_avg_opp_SUB_ATT  B_avg_REV  B_avg_opp_REV  \
0          0.500             1.0000       0.00            0.0   
1          1.500             0.0000       0.00            0.0   
2            NaN                NaN        NaN            NaN   
3            NaN                NaN        NaN            NaN   
4          0.125             0.1875       0.25            0.0   

   B_avg_SIG_STR_att  B_avg_SIG_STR_landed  B_avg_opp_SIG_STR_att  \
0            50.0000               20.0000                 84.000   
1            65.5000               35.0000                 50.000   
2                NaN                   NaN                    NaN   
3                NaN                   NaN                    NaN   
4           109.1875               57.9375                 50.625   

   B_avg_opp_SIG_STR_landed  B_avg_TOTAL_STR_att  B_avg_TOTAL_STR_landed  \
0                   45.0000              76.5000                  41.000   
1                   16.5000             113.5000                  68.500   
2                       NaN                  NaN                     NaN   
3                       NaN                  NaN                     NaN   
4                   28.4375             170.4375                 105.625   

   B_avg_opp_TOTAL_STR_att  B_avg_opp_TOTAL_STR_landed  B_avg_TD_att  \
0                  114.000                       64.00         1.500   
1                   68.500                       29.00         2.500   
2                      NaN                         NaN           NaN   
3                      NaN                         NaN           NaN   
4                   74.375                       44.25         5.375   

   B_avg_TD_landed  B_avg_opp_TD_att  B_avg_opp_TD_landed  B_avg_HEAD_att  \
0              1.0               9.0                6.500          39.500   
1              1.5               0.5                0.500          46.000   
2              NaN               NaN                  NaN             NaN   
3              NaN               NaN                  NaN             NaN   
4              1.5               2.0                0.625          77.375   

   B_avg_HEAD_landed  B_avg_opp_HEAD_att  B_avg_opp_HEAD_landed  \
0             11.000             63.0000                 27.500   
1             20.000             36.0000                  7.500   
2                NaN                 NaN                    NaN   
3                NaN                 NaN                    NaN   
4             31.375             41.5625                 22.625   

   B_avg_BODY_att  B_avg_BODY_landed  B_avg_opp_BODY_att  \
0          7.5000             7.0000             12.00

### Changing data types

In [15]:
# Check descriptive statistics for df_ufc dataframe
df_ufc.describe()

B_avg_KD  B_avg_opp_KD  B_avg_SIG_STR_pct  B_avg_opp_SIG_STR_pct  \
count  4585.000000   4585.000000        4585.000000            4585.000000   
mean      0.247476      0.176818           0.453310               0.434290   
std       0.378509      0.324633           0.130458               0.132618   
min       0.000000      0.000000           0.000000               0.000000   
25%       0.000000      0.000000           0.376489               0.351045   
50%       0.015625      0.000000           0.450000               0.427500   
75%       0.500000      0.250000           0.527500               0.510000   
max       5.000000      3.000000           1.000000               1.000000   

       B_avg_TD_pct  B_avg_opp_TD_pct  B_avg_SUB_ATT  B_avg_opp_SUB_ATT  \
count   4585.000000       4585.000000    4585.000000        4585.000000   
mean       0.292650          0.268742       0.478884           0.409276   
std        0.273628          0.267178       0.724229           0.653826   
min        0.000000          0.000000       0.000000           0.000000   
25%        0.031250          0.033203       0.000000           0.000000   
50%        0.250000          0.200000       0.148468           0.098389   
75%        0.500000          0.422812       0.750000           0.531250   
max        1.000000          1.000000       8.000000           7.000000   

        B_avg_REV  B_avg_opp_REV  B_avg_SIG_STR_att  B_avg_SIG_STR_landed  \
count  4585.00000    4585.000000        4585.000000           4585.000000   
mean      0.15673       0.141306          76.288358             33.337168   
std       0.34199       0.322623          51.175465             22.030474   
min       0.00000       0.000000           0.000000              0.000000   
25%       0.00000       0.000000          38.261230             17.000000   
50%       0.00000       0.000000          66.500000             29.687500   
75%       0.12500       0.125000         104.750000             45.500000   
max       4.00000       3.000000         456.000000            186.000000   

       B_avg_opp_SIG_STR_att  B_avg_opp_SIG_STR_landed  B_avg_TOTAL_STR_att  \
count            4585.000000               4585.000000          4585.000000   
mean               73.092895                 31.053464            97.635001   
std                49.911897                 21.371161            57.524779   
min                 0.000000                  0.000000             0.000000   
25%                36.405273                 15.562500            55.484375   
50%                64.312305                 27.343750            89.750000   
75%                99.875000                 42.000000           132.470703   
max               403.000000                202.000000           524.000000   

       B_avg_TOTAL_STR_landed  B_avg_opp_TOTAL_STR_att  \
count             4585.000000              4585.000000   
mean                52.281925                92.716135   
std                 31.771066                54.871771   
min                  0.000000                 0.000000   
25%                 28.937500                52.985352   
50%                 48.000000                85.531250   
75%                 70.000000               125.726562   
max                242.000000               433.000000   

       B_avg_opp_TOTAL_STR_landed  B_avg_TD_att  B_avg_TD_landed  \
count                 4585.000000   4585.000000      4585.000000   
mean                    48.504853      2.726783         1.109138   
std                     29.863232      2.836065         1.328047   
min                      0.000000      0.000000         0.000000   
25%                     26.875000      0.625000         0.062500   
50%                     44.411522      2.000000         0.750000   
75%                     65.187500      4.000000         1.562500   
max                    232.000000     19.000000        10.968750   

       B_avg_opp_TD_att  B_avg_opp_TD_landed  B_avg_HEAD_att  \
count       4585.000000  

#### No changes in data types.

### Transposing data

In [16]:
df_ufc.T

0                       1     \
red_fighter                             Adrian Yanez            Trevin Giles   
blue_fighter                           Gustavo Lopez           Roman Dolidze   
date                                      2021-03-20              2021-03-20   
location                      Las Vegas, Nevada, USA  Las Vegas, Nevada, USA   
winner                                           Red                     Red   
title_bout                                     False                   False   
weight_class                            Bantamweight            Middleweight   
B_avg_KD                                         0.0                     0.5   
B_avg_opp_KD                                     0.0                     0.0   
B_avg_SIG_STR_pct                               0.42                    0.66   
B_avg_opp_SIG_STR_pct                          0.495                   0.305   
B_avg_TD_pct                                    0.33                     0.3   
B_avg_opp_TD_pct                                0.36                     0.5   
B_avg_SUB_ATT                                    0.5                     1.5   
B_avg_opp_SUB_ATT                                1.0                     0.0   
B_avg_REV                                        0.0                     0.0   
B_avg_opp_REV                                    0.0                     0.0   
B_avg_SIG_STR_att                               50.0                    65.5   
B_avg_SIG_STR_landed                            20.0                    35.0   
B_avg_opp_SIG_STR_att                           84.0                    50.0   
B_avg_opp_SIG_STR_landed                        45.0                    16.5   
B_avg_TOTAL_STR_att                             76.5                   113.5   
B_avg_TOTAL_STR_landed                          41.0                    68.5   
B_avg_opp_TOTAL_STR_att                        114.0                    68.5   
B_avg_opp_TOTAL_STR_landed                      64.0                    29.0   
B_avg_TD_att                                     1.5                     2.5   
B_avg_TD_landed                                  1.0                     1.5   
B_avg_opp_TD_att                                 9.0                     0.5   
B_avg_opp_TD_landed                              6.5                     0.5   
B_avg_HEAD_att                                  39.5                    46.0   
B_avg_HEAD_landed                               11.0                    20.0   
B_avg_opp_HEAD_att                              63.0                    36.0   
B_avg_opp_HEAD_landed                           27.5                     7.5   
B_avg_BODY_att                                   7.5                    12.0   
B_avg_BODY_landed                                7.0                     8.0   
B_avg_opp_BODY_att                              12.0                     8.0   
B_avg_opp_BODY_landed                            9.0                     3.0   
B_avg_LEG_att                                    3.0                     7.5   
B_avg_LEG_landed                                 2.0                     7.0   
B_avg_opp_LEG_att                                9.0                     6.0   
B_avg_opp_LEG_landed                             8.5                     6.0   
B_avg_DISTANCE_att                              35.0                    58.0   
B_avg_DISTANCE_landed                           12.5                    30.0   
B_avg_opp_DISTANCE_att                          43.5                    48.0   
B_avg_opp_DISTANCE_landed                       17.5                    15.5   
B_avg_CLINCH_att                                10.5                     0.5   
B_avg_CLINCH_landed                              4.5                     0.5   
B_avg_opp_CLINCH_att                             4.0                     0.5   
B_avg_opp_CLINCH_landed                          3.0                     0.5   
B_avg_GROUND_att                                

### Turning the df_ufc dataframe from a wide format to a long format does not make a difference. Dataframe columns will remain the same as a wide format.

### Data consistency checks

### 01. Mixed-type data

### 02. Missing values

### 03. Duplicates

### Mixed-type data

In [17]:
df_ufc.describe()

B_avg_KD  B_avg_opp_KD  B_avg_SIG_STR_pct  B_avg_opp_SIG_STR_pct  \
count  4585.000000   4585.000000        4585.000000            4585.000000   
mean      0.247476      0.176818           0.453310               0.434290   
std       0.378509      0.324633           0.130458               0.132618   
min       0.000000      0.000000           0.000000               0.000000   
25%       0.000000      0.000000           0.376489               0.351045   
50%       0.015625      0.000000           0.450000               0.427500   
75%       0.500000      0.250000           0.527500               0.510000   
max       5.000000      3.000000           1.000000               1.000000   

       B_avg_TD_pct  B_avg_opp_TD_pct  B_avg_SUB_ATT  B_avg_opp_SUB_ATT  \
count   4585.000000       4585.000000    4585.000000        4585.000000   
mean       0.292650          0.268742       0.478884           0.409276   
std        0.273628          0.267178       0.724229           0.653826   
min        0.000000          0.000000       0.000000           0.000000   
25%        0.031250          0.033203       0.000000           0.000000   
50%        0.250000          0.200000       0.148468           0.098389   
75%        0.500000          0.422812       0.750000           0.531250   
max        1.000000          1.000000       8.000000           7.000000   

        B_avg_REV  B_avg_opp_REV  B_avg_SIG_STR_att  B_avg_SIG_STR_landed  \
count  4585.00000    4585.000000        4585.000000           4585.000000   
mean      0.15673       0.141306          76.288358             33.337168   
std       0.34199       0.322623          51.175465             22.030474   
min       0.00000       0.000000           0.000000              0.000000   
25%       0.00000       0.000000          38.261230             17.000000   
50%       0.00000       0.000000          66.500000             29.687500   
75%       0.12500       0.125000         104.750000             45.500000   
max       4.00000       3.000000         456.000000            186.000000   

       B_avg_opp_SIG_STR_att  B_avg_opp_SIG_STR_landed  B_avg_TOTAL_STR_att  \
count            4585.000000               4585.000000          4585.000000   
mean               73.092895                 31.053464            97.635001   
std                49.911897                 21.371161            57.524779   
min                 0.000000                  0.000000             0.000000   
25%                36.405273                 15.562500            55.484375   
50%                64.312305                 27.343750            89.750000   
75%                99.875000                 42.000000           132.470703   
max               403.000000                202.000000           524.000000   

       B_avg_TOTAL_STR_landed  B_avg_opp_TOTAL_STR_att  \
count             4585.000000              4585.000000   
mean                52.281925                92.716135   
std                 31.771066                54.871771   
min                  0.000000                 0.000000   
25%                 28.937500                52.985352   
50%                 48.000000                85.531250   
75%                 70.000000               125.726562   
max                242.000000               433.000000   

       B_avg_opp_TOTAL_STR_landed  B_avg_TD_att  B_avg_TD_landed  \
count                 4585.000000   4585.000000      4585.000000   
mean                    48.504853      2.726783         1.109138   
std                     29.863232      2.836065         1.328047   
min                      0.000000      0.000000         0.000000   
25%                     26.875000      0.625000         0.062500   
50%                     44.411522      2.000000         0.750000   
75%                     65.187500      4.000000         1.562500   
max                    232.000000     19.000000        10.968750   

       B_avg_opp_TD_att  B_avg_opp_TD_landed  B_avg_HEAD_att  \
count       4585.000000  

In [18]:
# check for mixed-type data in df_ufc dataframe
for col in df_ufc.columns.tolist():
  weird = (df_ufc[[col]].applymap(type) != df_ufc[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ufc[weird]) > 0:
    print (col)

B_Stance
R_Stance


In [19]:
df_ufc.head()

red_fighter       blue_fighter        date                location  \
0   Adrian Yanez      Gustavo Lopez  2021-03-20  Las Vegas, Nevada, USA   
1   Trevin Giles      Roman Dolidze  2021-03-20  Las Vegas, Nevada, USA   
2    Tai Tuivasa    Harry Hunsucker  2021-03-20  Las Vegas, Nevada, USA   
3  Cheyanne Buys  Montserrat Conejo  2021-03-20  Las Vegas, Nevada, USA   
4  Marion Reneau      Macy Chiasson  2021-03-20  Las Vegas, Nevada, USA   

  winner  title_bout       weight_class  B_avg_KD  B_avg_opp_KD  \
0    Red       False       Bantamweight     0.000           0.0   
1    Red       False       Middleweight     0.500           0.0   
2    Red       False        Heavyweight       NaN           NaN   
3   Blue       False   WomenStrawweight       NaN           NaN   
4   Blue       False  WomenBantamweight     0.125           0.0   

   B_avg_SIG_STR_pct  B_avg_opp_SIG_STR_pct  B_avg_TD_pct  B_avg_opp_TD_pct  \
0           0.420000                0.49500         0.330           0.36000   
1           0.660000                0.30500         0.300           0.50000   
2                NaN                    NaN           NaN               NaN   
3                NaN                    NaN           NaN               NaN   
4           0.535625                0.57875         0.185           0.16625   

   B_avg_SUB_ATT  B_avg_opp_SUB_ATT  B_avg_REV  B_avg_opp_REV  \
0          0.500             1.0000       0.00            0.0   
1          1.500             0.0000       0.00            0.0   
2            NaN                NaN        NaN            NaN   
3            NaN                NaN        NaN            NaN   
4          0.125             0.1875       0.25            0.0   

   B_avg_SIG_STR_att  B_avg_SIG_STR_landed  B_avg_opp_SIG_STR_att  \
0            50.0000               20.0000                 84.000   
1            65.5000               35.0000                 50.000   
2                NaN                   NaN                    NaN   
3                NaN                   NaN                    NaN   
4           109.1875               57.9375                 50.625   

   B_avg_opp_SIG_STR_landed  B_avg_TOTAL_STR_att  B_avg_TOTAL_STR_landed  \
0                   45.0000              76.5000                  41.000   
1                   16.5000             113.5000                  68.500   
2                       NaN                  NaN                     NaN   
3                       NaN                  NaN                     NaN   
4                   28.4375             170.4375                 105.625   

   B_avg_opp_TOTAL_STR_att  B_avg_opp_TOTAL_STR_landed  B_avg_TD_att  \
0                  114.000                       64.00         1.500   
1                   68.500                       29.00         2.500   
2                      NaN                         NaN           NaN   
3                      NaN                         NaN           NaN   
4                   74.375                       44.25         5.375   

   B_avg_TD_landed  B_avg_opp_TD_att  B_avg_opp_TD_landed  B_avg_HEAD_att  \
0              1.0               9.0                6.500          39.500   
1              1.5               0.5                0.500          46.000   
2              NaN               NaN                  NaN             NaN   
3              NaN               NaN                  NaN             NaN   
4              1.5               2.0                0.625          77.375   

   B_avg_HEAD_landed  B_avg_opp_HEAD_att  B_avg_opp_HEAD_landed  \
0             11.000             63.0000                 27.500   
1             20.000             36.0000                  7.500   
2                NaN                 NaN                    NaN   
3                NaN                 NaN                    NaN   
4             31.375             41.5625                 22.625   

   B_avg_BODY_att  B_avg_BODY_landed  B_avg_opp_BODY_att  \
0          7.5000             7.0000             12.00

### B_Stance and R_Stance contain mixed-type data.

In [20]:
# check datatypes of the df_ufc dataframe
df_ufc.dtypes

red_fighter                      object
blue_fighter                     object
date                             object
location                         object
winner                           object
title_bout                         bool
weight_class                     object
B_avg_KD                        float64
B_avg_opp_KD                    float64
B_avg_SIG_STR_pct               float64
B_avg_opp_SIG_STR_pct           float64
B_avg_TD_pct                    float64
B_avg_opp_TD_pct                float64
B_avg_SUB_ATT                   float64
B_avg_opp_SUB_ATT               float64
B_avg_REV                       float64
B_avg_opp_REV                   float64
B_avg_SIG_STR_att               float64
B_avg_SIG_STR_landed            float64
B_avg_opp_SIG_STR_att           float64
B_avg_opp_SIG_STR_landed        float64
B_avg_TOTAL_STR_att             float64
B_avg_TOTAL_STR_landed          float64
B_avg_opp_TOTAL_STR_att         float64
B_avg_opp_TOTAL_STR_landed      float64


#### B_Stance and R_Stance are already objects which is pandas' way of saying string.

### Missing values

In [21]:
# run a check for missing values
df_ufc.isnull().sum()

red_fighter                        0
blue_fighter                       0
date                               0
location                           0
winner                             0
title_bout                         0
weight_class                       0
B_avg_KD                        1427
B_avg_opp_KD                    1427
B_avg_SIG_STR_pct               1427
B_avg_opp_SIG_STR_pct           1427
B_avg_TD_pct                    1427
B_avg_opp_TD_pct                1427
B_avg_SUB_ATT                   1427
B_avg_opp_SUB_ATT               1427
B_avg_REV                       1427
B_avg_opp_REV                   1427
B_avg_SIG_STR_att               1427
B_avg_SIG_STR_landed            1427
B_avg_opp_SIG_STR_att           1427
B_avg_opp_SIG_STR_landed        1427
B_avg_TOTAL_STR_att             1427
B_avg_TOTAL_STR_landed          1427
B_avg_opp_TOTAL_STR_att         1427
B_avg_opp_TOTAL_STR_landed      1427
B_avg_TD_att                    1427
B_avg_TD_landed                 1427
B

#### Blue fighter stats have 1427 missing values and the red fighter stats have 712 missing values. These missing values can be explained because of empty statistics of new fighters joining UFC and fighting for the first time in the UFC. It is recommended to not impute these values.
#### Replace 'B_Stance' and 'R_Stance' NaN values by the mode.
#### Replace height, weight, and age NaN values by the median.

In [22]:
# view the missing values
df_stance = df_ufc[df_ufc['B_Stance'].isnull() == True]
df_stance

red_fighter             blue_fighter        date  \
58             Shana Dobson            Casey O'Neill  2021-02-20   
323         Yana Kunitskaya       Julija Stoliarenko  2020-08-08   
694             Khalid Taha              Bruno Silva  2019-10-05   
1147     Julija Stoliarenko              Leah Letson  2018-11-30   
1362            Ji Yeon Kim           Melinda Fabian  2018-06-23   
1613         DeAnna Bennett           Melinda Fabian  2017-12-01   
2240        Teruto Ishihara        Horacio Gutierrez  2016-08-06   
2286         Sage Northcutt            Enrique Marin  2016-07-09   
2296           Gray Maynard           Fernando Bruno  2016-07-08   
2307        Anthony Birchak             Dileno Lopes  2016-07-07   
2447           David Teymur          Martin Svensson  2016-02-27   
2537           Julian Erosa           Marcin Wrzosek  2015-12-11   
2546    Chris Gruetzemacher           Abner Lloveras  2015-12-11   
2558          Kailin Curran              Emily Kagan  2015-12-10   
2574          Erick Montano            Enrique Marin  2015-11-21   
2651          Joaquim Silva       Nazareno Malegarie  2015-09-05   
2681  Glaico Franca Moreira           Fernando Bruno  2015-08-01   
2684       Reginaldo Vieira             Dileno Lopes  2015-08-01   
2731           Kamaru Usman            Hayder Hassan  2015-07-12   
2832           Bec Rawlings               Lisa Ellis  2015-05-09   
2935           Daniel Kelly            Patrick Walsh  2015-02-14   
3010          Felice Herrig               Lisa Ellis  2014-12-12   
3052        Gabriel Benitez  Humberto Brown Morrison  2014-11-15   
3210         Cathal Pendred                Mike King  2014-07-19   
3230         Corey Anderson           Matt Van Buren  2014-07-06   
3370          Richard Walsh             Chris Indich  2014-04-16   
3526          Jessamyn Duke             Peggy Morgan  2013-11-30   
3712        Daniel Sarafian             Eddie Mendez  2013-06-08   
3786             Tor Troeng               Adam Cella  2013-04-06   
3873         Iuri Alcantara              Pedro Nobre  2013-01-19   
4078      Daron Cruickshank             Chris Tickle  2012-06-01   
4079             Myles Jury           Chris Saunders  2012-06-01   
4246          Bryan Caraway             Dustin Neace  2011-12-03   
4267             Aaron Rosa               Matt Lucas  2011-11-12   
4270             Terry Etim         Edward Faaloloto  2011-11-05   
4328              Ken Stone             Donny Walker  2011-09-17   
4332          TJ Waldburger              Mike Stumpf  2011-09-17   
4376          Jeff Hougland             Donny Walker  2011-07-02   
4388        Michael Johnson         Edward Faaloloto  2011-06-26   
4501            Waylon Lowe           Willamy Freire  2011-01-22   
4530          Cody McKenzie          Aaron Wilkinson  2010-12-04   
4576          Rob Broughton         Vinicius Queiroz  2010-10-16   
5785              Brad Gumm             CJ Fernandes  2000-09-22   
5789          Shonie Carter           Adrian Serrano  2000-06-09   
5791            Ian Freeman           Nate Schroeder  2000-06-09   
5818            Jens Pulver          Alfonso Alcarez  1999-09-24   
5843             Sione Latu             Joey Roberts  1999-03-05   
5847          Laverne Clark            Frank Caracci  1999-01-08   
5859         Tulio Palhares           Adriano Santos  1998-10-16   
5894            Alex Hunter               Sam Fulton  1997-07-27   
5903            Jack Nilson           Saeed Hosseini  1997-05-30   
5908        Jerry Bohlander               Nick Sanzo  1997-02-07   
5933            Geza Kalman            Dieusel Berto  1996-07-12   
5935          Steve Nelmark               Tai Bowden  1996-05-17   
5964           Scott Bessac               David Hood  1995-09-08   
5980               Jon Hess            Andy Anderson  1995-04-07   
5981             Guy Mezger               John Dowdy  1995-04-07   
5982          Dave Beneteau             Asbel Cancio  1995-04-07   
5988

In [23]:
# find the mode in 'B_stance' column
df_ufc['B_Stance'].mode()

0    Orthodox
Name: B_Stance, dtype: object

In [24]:
#fill NaNs with column mode in 'B_stance' column
df_ufc['B_Stance'].fillna('Orthodox', inplace = True)

In [25]:
# Check updated 'B_stance' column with no missing values
df_ufc['B_Stance'].isna().sum()

0

In [26]:
df_ufc.head(60)

red_fighter        blue_fighter        date  \
0            Adrian Yanez       Gustavo Lopez  2021-03-20   
1            Trevin Giles       Roman Dolidze  2021-03-20   
2             Tai Tuivasa     Harry Hunsucker  2021-03-20   
3           Cheyanne Buys   Montserrat Conejo  2021-03-20   
4           Marion Reneau       Macy Chiasson  2021-03-20   
5         Leonardo Santos        Grant Dawson  2021-03-20   
6              Song Kenan         Max Griffin  2021-03-20   
7           Derek Brunson       Kevin Holland  2021-03-20   
8          Montel Jackson       Jesse Strader  2021-03-20   
9             Bruno Silva             JP Buys  2021-03-20   
10             Manel Kape     Matheus Nicolau  2021-03-13   
11         Misha Cirkunov          Ryan Spann  2021-03-13   
12           Leon Edwards      Belal Muhammad  2021-03-13   
13       Charles Jourdain        Marcelo Rojo  2021-03-13   
14            Eryk Anders      Darren Stewart  2021-03-13   
15            Angela Hill        Ashley Yoder  2021-03-13   
16      Jonathan Martinez         Davey Grant  2021-03-13   
17                Dan Ige        Gavin Tucker  2021-03-13   
18       Nasrat Haqparast         Rafa Garcia  2021-03-13   
19          Cortney Casey          JJ Aldrich  2021-03-13   
20   Matthew Semelsberger          Jason Witt  2021-03-13   
21             Rani Yahya       Ray Rodriguez  2021-03-13   
22        Gloria de Paula        Jinh Yu Frey  2021-03-13   
23          Dominick Cruz        Casey Kenney  2021-03-06   
24            Song Yadong      Kyler Phillips  2021-03-06   
25               Petr Yan   Aljamain Sterling  2021-03-06   
26         Jan Blachowicz     Israel Adesanya  2021-03-06   
27          Thiago Santos    Aleksandar Rakic  2021-03-06   
28       Joseph Benavidez       Askar Askarov  2021-03-06   
29           Amanda Nunes      Megan Anderson  2021-03-06   
30        Islam Makhachev          Drew Dober  2021-03-06   
31            Tim Elliott     Jordan Espinosa  2021-03-06   
32          Livinha Souza        Amanda Lemos  2021-03-06   
33       Rogerio Bontorin     Kai Kara-France  2021-03-06   
34             Uros Medic          Aalon Cruz  2021-03-06   
35      Kennedy Nzechukwu       Carlos Ulberg  2021-03-06   
36             Sean Brady       Jake Matthews  2021-03-06   
37         Mario Bautista        Trevin Jones  2021-03-06   
38  Jairzinho Rozenstruik          Ciryl Gane  2021-02-27   
39     Montana De La Rosa   Mayra Bueno Silva  2021-02-27   
40          Vince Cachero     Ronnie Lawrence  2021-02-27   
41    Alexander Hernandez       Thiago Moises  2021-02-27   
42           Pedro Munhoz       Jimmie Rivera  2021-02-27   
43           Alex Caceres         Kevin Croom  2021-02-27   
44           Alexis Davis         Sabina Mazo  2021-02-27   
45          Nikita Krylov    Magomed Ankalaev  2021-02-27   
46          Dustin Jacoby       Maxim Grishin  2021-02-27   
47        Aleksei Oleinik       Chris Daukaus  2021-02-20   
48           Charles Rosa      Darrick Minner  2021-02-20   
49             Phil Hawes  Nassourdine Imavov  2021-02-20   
50         Curtis Blaydes       Derrick Lewis  2021-02-20   
51         Eddie Wineland      John Castaneda  2021-02-20   
52        Andrei Arlovski        Tom Aspinall  2021-02-20   
53           Jared Gordon        Danny Chavez  2021-02-20   
54          Ketlen Vieira     Yana Kunitskaya  2021-02-20   
55         Aiemann Zahabi     Drako Rodriguez  2021-02-20   
56         Serghei Spivac      Jared Vanderaa  2021-02-20   
57          Nate Landwehr        Julian Erosa  2021-02-20   
58           Shana Dobson       Casey O'Neill  2021-02-20   
59          Maycee Barber        Alexa Grasso  2021-02-13   

                  location winner  title_bout        weight_class  B_avg_KD  \
0   Las Vegas, Nevada, USA    Red       False        Bantamweight  0.000000   
1   Las Vegas, Nevada, USA    Red       False        Middleweight  0.500000   
2   Las Vegas, Nevada, USA    Red       False         H

In [27]:
# find the mode in 'R_stance' column
df_ufc['R_Stance'].mode()

0    Orthodox
Name: R_Stance, dtype: object

In [28]:
#fill NaNs with column mode in 'R_stance' column
df_ufc['R_Stance'].fillna('Orthodox', inplace = True)

In [29]:
# Check updated 'R_stance' column with no missing values
df_ufc['R_Stance'].isna().sum()

0

In [30]:
df_ufc.shape

(6012, 143)

In [33]:
# find the median in 'B_Height_cms' column
df_ufc['B_Height_cms'].median()

180.34

In [34]:
# fill NaNs with column mode in 'B_Height_cms' column
df_ufc['B_Height_cms'].fillna('180.34', inplace = True)

In [35]:
# Check updated 'B_Height_cms' column with no missing values
df_ufc['B_Height_cms'].isna().sum()

0

In [36]:
# find the median in 'R_Height_cms' column
df_ufc['R_Height_cms'].median()

180.34

In [37]:
#fill NaNs with column mode in 'R_Height_cms' column
df_ufc['R_Height_cms'].fillna('180.34', inplace = True)

In [38]:
# Check updated 'R_Height_cms' column with no missing values
df_ufc['R_Height_cms'].isna().sum()

0

In [39]:
# find the median in 'B_Weight_lbs' column
df_ufc['B_Weight_lbs'].median()

170.0

In [41]:
# fill NaNs with column mode in 'B_Weight_lbs' column
df_ufc['B_Weight_lbs'].fillna('170.0', inplace = True)

In [42]:
# Check updated 'B_Weight_lbs' column with no missing values
df_ufc['B_Weight_lbs'].isna().sum()

0

In [43]:
# find the median in 'R_Weight_lbs' column
df_ufc['R_Weight_lbs'].median()

170.0

In [45]:
# fill NaNs with column mode in 'R_Weight_lbs' column
df_ufc['R_Weight_lbs'].fillna('170.0', inplace = True)

In [46]:
# Check updated 'R_Weight_lbs' column with no missing values
df_ufc['R_Weight_lbs'].isna().sum()

0

In [47]:
# find the median in 'B_age' column
df_ufc['B_age'].median()

29.0

In [49]:
# fill NaNs with column mode in 'B_age' column
df_ufc['B_age'].fillna('29.0', inplace = True)

In [50]:
# Check updated 'B_age' column with no missing values
df_ufc['B_age'].isna().sum()

0

In [51]:
# find the median in 'R_age' column
df_ufc['R_age'].median()

29.0

In [53]:
# fill NaNs with column mode in 'R_age' column
df_ufc['R_age'].fillna('29.0', inplace = True)

In [54]:
# Check updated 'R_age' column with no missing values
df_ufc['R_age'].isna().sum()

0

In [55]:
# print the frequency of the 'location' column
df_ufc['location'].value_counts(dropna = False)

Las Vegas, Nevada, USA                         1565
Abu Dhabi, Abu Dhabi, United Arab Emirates      157
London, England, United Kingdom                 114
Newark, New Jersey, USA                          90
Anaheim, California, USA                         84
Montreal, Quebec, Canada                         81
Chicago, Illinois, USA                           81
Atlantic City, New Jersey, USA                   80
Los Angeles, California, USA                     79
Toronto, Ontario, Canada                         74
Boston, Massachusetts, USA                       72
Stockholm, Sweden                                72
Rio de Janeiro, Brazil                           69
Houston, Texas, USA                              69
Denver, Colorado, USA                            66
Sao Paulo, Sao Paulo, Brazil                     63
Dallas, Texas, USA                               63
Sacramento, California, USA                      62
Nashville, Tennessee, USA                        59
Mexico City,

In [56]:
# create 'country' column using for-loops function
result = []

for value in df_ufc ["location"]:
    if value in ['Albany, New York, USA', 'Albuquerque, New Mexico, USA', 'Anaheim, California, USA', 'Atlanta, Georgia, USA', 'Atlantic City, New Jersey, USA', 'Auburn Hills, Michigan, USA', 'Augusta, Georgia, USA', 'Austin, Texas, USA', 'Baltimore, Maryland, USA', 'Bangor, Maine, USA', 'Bay St. Louis, Mississippi, USA', 'Birmingham, Alabama, USA', 'Boise, Idaho, USA', 'Bossier City, Louisiana, USA', 'Boston, Massachusetts, USA', 'Brooklyn, New York, USA', 'Broomfield, Colorado, USA', 'Buffalo, New York, USA', 'Casper, Wyoming, USA', 'Cedar Rapids, Iowa, USA', 'Charlotte, North Carolina, USA', 'Chicago, Illinois, USA', 'Cincinnati, Ohio, USA', 'Cleveland, Ohio, USA', 'Columbus, Ohio, USA', 'Dallas, Texas, USA', 'Denver, Colorado, USA', 'Detroit, Michigan, USA', 'Dothan, Alabama, USA', 'Duluth, Georgia, USA', 'East Rutherford, New Jersey, USA', 'Fairfax, Virginia, USA', 'Fayetteville, North Carolina, USA', 'Fort Campbell, Kentucky, USA', 'Fresno, California, USA', 'Glendale, Arizona, USA', 'Greenville, South Carolina, USA', 'Hidalgo, Texas, USA', 'Hollywood, Florida, USA', 'Houston, Texas, USA', 'Indianapolis, Indiana, USA', 'Jacksonville, Florida, USA', 'Kansas City, Missouri, USA', 'Killeen, Texas, USA', 'Lake Charles, Louisiana, USA', 'Las Vegas, Nevada, USA', 'Ledyard, Connecticut, USA', 'Lincoln, Nebraska, USA', 'Los Angeles, California, USA', 'Louisville, Kentucky, USA', 'Memphis, Tennessee, USA', 'Miami, Florida, USA', 'Milwaukee, Wisconsin, USA', 'Minneapolis, Minnesota, USA', 'Mobile, Alabama, USA', 'Nashville, Tennessee, USA', 'New Orleans, Louisiana, USA', 'New York City, New York, USA', 'Newark, New Jersey, USA', 'Norfolk, Virginia, USA', 'Oakland, California, USA', 'Oklahoma City, Oklahoma, USA', 'Omaha, Nebraska, USA', 'Orlando, Florida, USA', 'Philadelphia, Pennsylvania, USA', 'Phoenix, Arizona, USA', 'Pittsburgh, Pennsylvania, USA', 'Portland, Oregon, USA', 'Raleigh, North Carolina, USA', 'Rio Rancho, New Mexico, USA', 'Rochester, New York, USA', 'Rosemont, Illinois, USA', 'Sacramento, California, USA', 'Salt Lake City, Utah, USA', 'San Antonio, Texas, USA', 'San Diego, California, USA', 'San Jose, California, USA', 'Seattle, Washington, USA', 'Sioux Falls, South Dakota, USA', 'St. Louis, Missouri, USA', 'Sunrise, Florida, USA', 'Tampa, Florida, USA', 'Tulsa, Oklahoma, USA', 'Uncasville, Connecticut, USA', 'Uniondale, New York, USA', 'Utica, New York, USA', 'Washington, DC, USA', 'Wichita, Kansas, USA']:
        result.append("United States")
    elif value in ['Abu Dhabi, Abu Dhabi, United Arab Emirates', 'Abu Dhabi, United Arab Emirates']:
        result.append("United Arab Emirates")
    elif value in ['Melbourne, Victoria, Australia', 'Adelaide, South Australia, Australia', 'Perth, Western Australia, Australia', 'Sydney, New South Wales, Australia', 'Melbourne, Victoria, Australia', 'Brisbane, Queensland, Australia', 'Gold Coast, Queensland, Australia']:
        result.append("Australia")
    elif value in ['Brasilia, Distrito Federal, Brazil', 'Sao Paulo, Sao Paulo, Brazil', 'Rio de Janeiro, Rio de Janeiro, Brazil', 'Fortaleza, Ceara, Brazil', 'Belem, Para, Brazil', 'Fortaleza, Ceara, Brazil', 'Curitiba, Parana, Brazil', 'Goiania, Goias, Brazil', 'Rio de Janeiro, Brazil', 'Porto Alegre, Rio Grande do Sul, Brazil', 'Barueri, Sao Paulo, Brazil', 'Uberlandia, Minas Gerais, Brazil', 'Sao Paulo, Brazil', 'Natal, Rio Grande do Norte, Brazil', 'Jaragua do Sul, Santa Catarina, Brazil', 'Belo Horizonte, Minas Gerais, Brazil']:
        result.append("Brazil")
    elif value in ['Shenzhen, Guangdong, China', 'Beijing, China', 'Shanghai, Hebei, China', 'Macau, China']:
        result.append("China")
    elif value in ['London, England, United Kingdom', 'Liverpool, England, United Kingdom', 'Glasgow, Scotland, United Kingdom', 'Belfast, Northern Ireland, United Kingdom', 'Manchester, England, United Kingdom', 'Nottingham, England, United Kingdom', 'Birmingham, England, United Kingdom', 'Newcastle, England, United Kingdom' ]:
        result.append("United Kingdom")
    elif value in ['Hamburg, Hamburg, Germany', 'Berlin, Berlin, Germany', 'Berlin, Germany', 'Oberhausen, North Rhine-Westphalia, Germany', 'Cologne, North Rhine-Westphalia, Germany']:
        result.append("Germany")
    elif value in ['Copenhagen, Denmark']:
        result.append("Denmark")
    elif value in ['Dublin, Leinster, Ireland']:
        result.append("Ireland")
    elif value in ['Auckland, New Zealand']:
        result.append("New Zealand")
    elif value in ['Buenos Aires, Buenos Aires, Argentina']:
        result.append("Argentina")
    elif value in ['Busan, South Korea', 'Seoul, South Korea']:
        result.append("South Korea")    
    elif value in ['Vancouver, British Columbia, Canada', 'Edmonton, Alberta, Canada', 'Ottawa, Ontario, Canada', 'Toronto, Ontario, Canada', 'Moncton, New Brunswick, Canada', 'Calgary, Alberta, Canada', 'Winnipeg, Manitoba, Canada', 'Halifax, Nova Scotia, Canada', 'Saskatoon, Saskatchewan, Canada', 'Montreal, Quebec, Canada', 'Quebec City, Quebec, Canada']:
        result.append("Canada")    
    elif value in ['Gdansk, Poland', 'Krakow, Poland']:
        result.append("Poland")
    elif value in ['Mexico City, Distrito Federal, Mexico', 'Monterrey, Nuevo Leon, Mexico', 'Mexico City, Distrito Federal, Mexico']:
        result.append("Mexico")                  
    elif value in ['Saitama, Saitama, Japan', 'Saitama, Japan', 'Tokyo, Japan', 'Chiba, Japan', 'Yokohama, Kanagawa, Japan']:
        result.append("Japan")
    elif value in ['Zagreb, Croatia']:
        result.append("Croatia")                  
    elif value in ['San Juan, Puerto Rico']:
        result.append("Puerto Rico")
    elif value in ['Santiago, Chile']:
        result.append("Chile")
    elif value in ['Kallang, Singapore', 'Singapore, Singapore']:
        result.append("Singapore")                  
    elif value in ['Montevideo, Uruguay']:
        result.append("Uruguay")
    elif value in ['Moscow, Moscow, Russia', 'Saint Petersburg, Saint Petersburg, Russia']:
        result.append("Russia")                  
    elif value in ['Prague, Czech Republic']:
        result.append("Czech Republic")
    elif value in ['Rotterdam, Zuid-Holland, Netherlands']:
        result.append("Netherlands")
    elif value in ['Manila, Philippines']:
        result.append("Philippines")
    elif value in ['Stockholm, Sweden']:
        result.append("Sweden")

In [57]:
result

['United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United States',
 'United S

In [58]:
df_ufc['country'] = result

In [59]:
df_ufc.head(20)

red_fighter       blue_fighter        date                location  \
0        Adrian Yanez      Gustavo Lopez  2021-03-20  Las Vegas, Nevada, USA   
1        Trevin Giles      Roman Dolidze  2021-03-20  Las Vegas, Nevada, USA   
2         Tai Tuivasa    Harry Hunsucker  2021-03-20  Las Vegas, Nevada, USA   
3       Cheyanne Buys  Montserrat Conejo  2021-03-20  Las Vegas, Nevada, USA   
4       Marion Reneau      Macy Chiasson  2021-03-20  Las Vegas, Nevada, USA   
5     Leonardo Santos       Grant Dawson  2021-03-20  Las Vegas, Nevada, USA   
6          Song Kenan        Max Griffin  2021-03-20  Las Vegas, Nevada, USA   
7       Derek Brunson      Kevin Holland  2021-03-20  Las Vegas, Nevada, USA   
8      Montel Jackson      Jesse Strader  2021-03-20  Las Vegas, Nevada, USA   
9         Bruno Silva            JP Buys  2021-03-20  Las Vegas, Nevada, USA   
10         Manel Kape    Matheus Nicolau  2021-03-13  Las Vegas, Nevada, USA   
11     Misha Cirkunov         Ryan Spann  2021-03-13  Las Vegas, Nevada, USA   
12       Leon Edwards     Belal Muhammad  2021-03-13  Las Vegas, Nevada, USA   
13   Charles Jourdain       Marcelo Rojo  2021-03-13  Las Vegas, Nevada, USA   
14        Eryk Anders     Darren Stewart  2021-03-13  Las Vegas, Nevada, USA   
15        Angela Hill       Ashley Yoder  2021-03-13  Las Vegas, Nevada, USA   
16  Jonathan Martinez        Davey Grant  2021-03-13  Las Vegas, Nevada, USA   
17            Dan Ige       Gavin Tucker  2021-03-13  Las Vegas, Nevada, USA   
18   Nasrat Haqparast        Rafa Garcia  2021-03-13  Las Vegas, Nevada, USA   
19      Cortney Casey         JJ Aldrich  2021-03-13  Las Vegas, Nevada, USA   

   winner  title_bout       weight_class  B_avg_KD  B_avg_opp_KD  \
0     Red       False       Bantamweight  0.000000      0.000000   
1     Red       False       Middleweight  0.500000      0.000000   
2     Red       False        Heavyweight       NaN           NaN   
3    Blue       False   WomenStrawweight       NaN           NaN   
4    Blue       False  WomenBantamweight  0.125000      0.000000   
5    Blue       False        Lightweight  0.000000      0.000000   
6    Blue       False       Welterweight  0.046875      0.125000   
7     Red       False       Middleweight  0.156250      0.000000   
8     Red       False       Bantamweight       NaN           NaN   
9     Red       False          Flyweight       NaN           NaN   
10   Blue       False          Flyweight  1.000000      0.625000   
11   Blue       False   LightHeavyweight  1.062500      0.000000   
12   Draw       False       Welterweight  0.000000      0.064941   
13    Red       False      Featherweight       NaN           NaN   
14   Draw       False       Middleweight  0.023438      0.000000   
15    Red       False   WomenStrawweight  0.015625      0.000000   
16   Blue       False       Bantamweight  0.500000      0.625000   
17    Red       False      Featherweight  0.000000      0.312500   
18    Red       False        Lightweight       NaN           NaN   
19   Blue       False     WomenFlyweight  0.000000      0.000000   

    B_avg_SIG_STR_pct  B_avg_opp_SIG_STR_pct  B_avg_TD_pct  B_avg_opp_TD_pct  \
0            0.420000               0.495000      0.330000          0.360000   
1            0.660000               0.305000      0.300000          0.500000   
2                 NaN                    NaN           NaN               NaN   
3                 NaN                    NaN           NaN               NaN   
4            0.535625               0.578750      0.185000          0.166250   
5            0.515000               0.473750      0.435000          0.250000   
6            0.459277               0.404687      0.322188          0.068867   
7            0.791582               0.850586      0.269199          0.393398   
8                 NaN                    NaN           NaN               NaN   
9                 NaN                    NaN           NaN               NaN   
10           0.47375

In [60]:
df_ufc.shape

(6012, 144)

### Duplicates

In [61]:
# run a check for duplicate values
df_ufc_dups = df_ufc[df_ufc.duplicated()]
df_ufc_dups

Empty DataFrame
Columns: [red_fighter, blue_fighter, date, location, winner, title_bout, weight_class, B_avg_KD, B_avg_opp_KD, B_avg_SIG_STR_pct, B_avg_opp_SIG_STR_pct, B_avg_TD_pct, B_avg_opp_TD_pct, B_avg_SUB_ATT, B_avg_opp_SUB_ATT, B_avg_REV, B_avg_opp_REV, B_avg_SIG_STR_att, B_avg_SIG_STR_landed, B_avg_opp_SIG_STR_att, B_avg_opp_SIG_STR_landed, B_avg_TOTAL_STR_att, B_avg_TOTAL_STR_landed, B_avg_opp_TOTAL_STR_att, B_avg_opp_TOTAL_STR_landed, B_avg_TD_att, B_avg_TD_landed, B_avg_opp_TD_att, B_avg_opp_TD_landed, B_avg_HEAD_att, B_avg_HEAD_landed, B_avg_opp_HEAD_att, B_avg_opp_HEAD_landed, B_avg_BODY_att, B_avg_BODY_landed, B_avg_opp_BODY_att, B_avg_opp_BODY_landed, B_avg_LEG_att, B_avg_LEG_landed, B_avg_opp_LEG_att, B_avg_opp_LEG_landed, B_avg_DISTANCE_att, B_avg_DISTANCE_landed, B_avg_opp_DISTANCE_att, B_avg_opp_DISTANCE_landed, B_avg_CLINCH_att, B_avg_CLINCH_landed, B_avg_opp_CLINCH_att, B_avg_opp_CLINCH_landed, B_avg_GROUND_att, B_avg_GROUND_landed, B_avg_opp_GROUND_att, B_avg_opp_GROUND_landed, B_avg_CTRL_time(seconds), B_avg_opp_CTRL_time(seconds), B_total_time_fought(seconds), B_total_rounds_fought, B_total_title_bouts, B_current_win_streak, B_current_lose_streak, B_longest_win_streak, B_wins, B_losses, B_draw, B_win_by_Decision_Majority, B_win_by_Decision_Split, B_win_by_Decision_Unanimous, B_win_by_KO/TKO, B_win_by_Submission, B_win_by_TKO_Doctor_Stoppage, B_Stance, B_Height_cms, B_Reach_cms, B_Weight_lbs, R_avg_KD, R_avg_opp_KD, R_avg_SIG_STR_pct, R_avg_opp_SIG_STR_pct, R_avg_TD_pct, R_avg_opp_TD_pct, R_avg_SUB_ATT, R_avg_opp_SUB_ATT, R_avg_REV, R_avg_opp_REV, R_avg_SIG_STR_att, R_avg_SIG_STR_landed, R_avg_opp_SIG_STR_att, R_avg_opp_SIG_STR_landed, R_avg_TOTAL_STR_att, R_avg_TOTAL_STR_landed, R_avg_opp_TOTAL_STR_att, R_avg_opp_TOTAL_STR_landed, R_avg_TD_att, R_avg_TD_landed, R_avg_opp_TD_att, R_avg_opp_TD_landed, R_avg_HEAD_att, R_avg_HEAD_landed, R_avg_opp_HEAD_att, R_avg_opp_HEAD_landed, ...]
Index: []

#### There are no duplicates in the df_ufc dataframe.

In [62]:
# Export df_ufc dataframe in "Prepared Data"
df_ufc.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_ufc_clean.csv'))